eradiate/radprops/rad_profile.py

```diff
@@ -559,105 +512,66 @@ def default_absorption_data_set(wavelength):
+        """
+        wavelength = spectral_ctx.wavelength
+        profile = self.eval_thermoprops_profile()
+
         if self.absorption_data_set is None:  # ! this is never tested
             data_set = self.default_absorption_data_set(wavelength=wavelength)
         else:
             data_set = xr.open_dataset(self.absorption_data_set)
+
+        # Compute scattering coefficient
         return compute_sigma_a(
             ds=data_set,
             wl=wavelength,
-            p=self._thermoprops.p.values,
-            n=self._thermoprops.n.values,
+            p=profile.p.values,
+            n=profile.n.values,
             fill_values=dict(pt=0.0),  # us76_u86_4 dataset is limited to pressures above
             # 0.101325 Pa, but us76 thermophysical profile goes below that
             # value for altitudes larger than 93 km. At these altitudes, the
             # number density is so small compared to that at the sea level that
             # we assume it is negligible.
-        ).to("km^-1")
+        )
 
-    def compute_sigma_s(self):
+    def eval_sigma_s(self, spectral_ctx):
         """
-        Compute scattering coefficient.
+        Evaluate scattering coefficient given spectral context.
         """
-        mode = eradiate.mode()
-        if not mode.is_monochromatic():
-            raise ModeError(f"unsupported mode {mode.id}")
-        else:
-            wavelength = mode.wavelength
-            self._wavelength = wavelength
-        profile = self._thermoprops
+        wavelength = spectral_ctx.wavelength
+        profile = self.eval_thermoprops_profile()
+
         return compute_sigma_s_air(
             wavelength=wavelength,
             number_density=ureg.Quantity(profile.n.values, profile.n.units),
-        ).to("km^-1")
-
-    def compute_thermoprops(self):
-        """
-        Compute the atmosphere thermophysical properties.
-        """
-        return us76.make_profile(self.levels)
-
-    def update(self):
-        """
-        Update object's internal state.
-        """
-        self._thermoprops = self.compute_thermoprops()
-        self._sigma_a_values = self.compute_sigma_a()
-        self._sigma_s_values = self.compute_sigma_s()
-
-    @property
-    def albedo(self):
-        """
-        Return albedo.
-
-        Returns → :class:`pint.Quantity`:
-            Profile albedo.
-        """
-        return (self.sigma_s / self.sigma_t).to(ureg.dimensionless)
-
-    @property
-    def sigma_a(self):
-        """
-        Return absorption coefficient.
-
-        Returns → :class:`pint.Quantity`:
-            Profile absorption coefficient.
-        """
-        return self._sigma_a_values[np.newaxis, np.newaxis, ...]
+        )
 
-    @property
-    def sigma_s(self):
-        """
-        Return scattering coefficient.
+    def albedo(self, spectral_ctx=None):
+        return (self.sigma_s(spectral_ctx) / self.sigma_t(spectral_ctx)).to(
+            ureg.dimensionless
+        )
```

Comment:
If I'm not mistaken, calling `albedo()` results in three calls to `eval_thermoprops_profile()`; is the result of `eval_thermoprops_profile()` cached then, or is the thermophysical properties profile computed three times?

Reply:
It is evaluated 3 times. I checked: This results in a perceptible performance penalty when running tests (the `not slow` series goes from 9 s to 12 s on my machine). There is little I can do for now, though: this requires careful examination because handling cache correctly is not trivial (we have plenty of mutable data, which greatly complicates cache management by making pretty much all caching libs unusable).